ChiModules/CFEMDiffusion/cfem_diffusion_solver.cc

```diff
@@ -0,0 +1,337 @@
+ 
+  chi::log.Log() << "Num local DOFs: " << num_local_dofs;
+  chi::log.Log() << "Num globl DOFs: " << num_globl_dofs;
+
+  //============================================= Initializes Mats and Vecs
+  const auto n = static_cast<int64_t>(num_local_dofs);
+  const auto N = static_cast<int64_t>(num_globl_dofs);
+ 
+  A = chi_math::PETScUtils::CreateSquareMatrix(n,N);
+  x = chi_math::PETScUtils::CreateVector(n,N);
+  b = chi_math::PETScUtils::CreateVector(n,N);
+ 
+  std::vector<int64_t> nodal_nnz_in_diag;
+  std::vector<int64_t> nodal_nnz_off_diag;
+  sdm.BuildSparsityPattern(nodal_nnz_in_diag,nodal_nnz_off_diag, OneDofPerNode);
+ 
+  chi_math::PETScUtils::InitMatrixSparsity(A,
+                                           nodal_nnz_in_diag,
+                                           nodal_nnz_off_diag);  
+  if (field_functions.empty())
+  {
+    auto unk_man = OneDofPerNode;
+    auto initial_field_function =
+      std::make_shared<chi_physics::FieldFunction>(
+        std::string("phi"),   //Text name
+        sdm_ptr,              //Spatial Discretization
+        &x,                   //Data vector
+        unk_man);             //Unknown Manager
+
+      field_functions.push_back(initial_field_function);
+      chi::fieldfunc_stack.push_back(initial_field_function);
+  }//if not ff set
+
+}//end initialize
+
+//========================================================== Execute
+void cfem_diffusion::Solver::Execute()
+{
+  chi::log.Log() << "\nExecuting CFEM Diffusion solver";
+
+  const auto& grid = *grid_ptr;
+  const auto& sdm  = *sdm_ptr;
+
+  lua_State* L = chi::console.consoleState;
+
+  //============================================= Assemble the system
+  chi::log.Log() << "Assembling system: ";
+  for (const auto& cell : grid.local_cells)
+  {
+    const auto& cell_mapping = sdm.GetCellMapping(cell);
+    const auto  qp_data      = cell_mapping.MakeVolumeQuadraturePointData();
+ 
+    const auto imat  = cell.material_id;
+    const size_t num_nodes = cell_mapping.NumNodes();
+    MatDbl Acell(num_nodes, VecDbl(num_nodes, 0.0));
+    VecDbl cell_rhs(num_nodes, 0.0);
+ 
+    for (size_t i=0; i<num_nodes; ++i)
+    {
+      for (size_t j=0; j<num_nodes; ++j)
+      {
+        double entry_aij = 0.0;
+        for (size_t qp : qp_data.QuadraturePointIndices())
+        {
+          entry_aij +=
+            (
+              CallLua_iXYZFunction(L, "D_coef",imat,qp_data.QPointXYZ(qp))  *
+              qp_data.ShapeGrad(i, qp).Dot(qp_data.ShapeGrad(j, qp))
+              +
+              CallLua_iXYZFunction(L, "Sigma_a",imat,qp_data.QPointXYZ(qp))  *
+              qp_data.ShapeValue(i, qp) * qp_data.ShapeValue(j, qp)
+            )
+            *
+            qp_data.JxW(qp);
+        }//for qp
+        Acell[i][j] = entry_aij;
+      }//for j
+      for (size_t qp : qp_data.QuadraturePointIndices())
+        cell_rhs[i] += CallLua_iXYZFunction(L,"Q_ext",imat,qp_data.QPointXYZ(qp)) * qp_data.ShapeValue(i, qp) * qp_data.JxW(qp);
```

Comment:
To be discussed. The names of the D, sigma and Q functions are hardcoded here as `D_coef`, `Sigma_a` and `Q_ext`. These names are fine as defaults but we should have the option to change them.

Consider adding them as `BasicOptions`. This will involve adding them to the constructor:
```c++
cfem_diffusion::Solver::Solver(const std::string& in_solver_name):
  chi_physics::Solver(in_solver_name, { {"max_iters", int64_t(500)   },
                                        {"residual_tolerance", 1.0e-2},
                                        {"D_function",std::string("D_coeff")},
                                        {"Sigma_function", std::string("Sigma_a")},
                                        {"Q_function", std::string("Q_ext")}})
{}
```
Then getting them at the top of the Execute routine with:
```c++
const auto D_func_name = basic_options("D_function").StringValue();
```

Users can then automatically set it via lua with something like
```lua
chiSolverSetBasicOption(phys1,"D_function","SexySolverDFunc")
```